src/Tokenizer/Tokens.php

```diff
@@ -877,18 +877,23 @@ public function insertAt($index, $items)
 
         $oldSize = \count($this);
         $this->changed = true;
+        $this->blockEndCache = [];
         $this->setSize($oldSize + $itemsCnt);
 
+        // since we only move already existing items around, we directly call into SplFixedArray::offset* methods.
+        // that way we get around additional overhead this class adds with overridden offset* methods.
         for ($i = $oldSize + $itemsCnt - 1; $i >= $index; --$i) {
-            $this[$i] = $this->offsetExists($i - $itemsCnt) ? $this[$i - $itemsCnt] : new Token('');
+            $oldItem = parent::offsetExists($i - $itemsCnt) ? parent::offsetGet($i - $itemsCnt) : new Token('');
```

Comment:
I tried to use parent::offsetGet() before offsetExists and determine property existance based on offsetGet returning NULL. 
Goal: reduce the number of offsetExists() calls

But this didn‘t worked out - not sure why.
Since this changes already is a major perf improvement over the current code, we should check that again in a separate PR